### DISCUSSION TITLE
kyuubi supports spark engine multi-mode at the same time

### DISCUSSION THREAD

**Discussion by 942011334** [apache/kyuubi#1580] [Q&A] [answered]
Can kyuubi support spark engine multi-mode at the same time？For example, when we use CONNECTION mode, we also use USER mode in a kyuubi server.

**Comment by 942011334** [comment]
@yaooqinn

**Comment by yaooqinn** [comment, marked as answer]
yes, it can be configured in each connection url.

**Reply by yaooqinn** [reply]
For further reading,

https://mp.weixin.qq.com/s/gi8w2oCsXqAiXn9blv9BBw
https://kyuubi.apache.org/docs/r1.4.0-incubating/deployment/engine_share_level.html#hybrid

**Comment by pan3793** [comment]
Yes, you can overwrite the specific conf in connection URL. e.g.
```
beeline -n batch_user -u 'jdbc:hive2://kyuubi:10009/;?spark.driver.memory=8G#kyuubi.engine.share.level=CONNECTION'
```

**Reply by kmsantanu** [reply]
which version of the kyuubi server it is supportin?

**Reply by pan3793** [reply]
Any version after 1.x, we recommend to use the latest version.

**Comment by kmsantanu** [comment]
how to solve our problem in production
problem 1:
1. We have deploy kyubi on machine-1 and we want two different users can connect kyuubi  engine with same spark session.
2. We have one user and this user has two task which is running paralle .How kyubi can run the parallel request like count and groupby operation from a same user.
3. we are using cluster mode spark deployment like .one master node and 4 slave node . 
Each slave has 80gb ram and 24core.Now one kyubi engine is taking all resources and I can not start my other spark application.what will optimise kyuubi configuration for production to run multiple user with limited resources

**Reply by pan3793** [reply]
1.  I guess you mean two user share the "same spark driver", you can use SERVER/GROUP engine share level.
2. Open two beeline connection, then you can submit query parallel, as long as you have enough executor to schedule the tasks.
3. You can leverage spark dynamic allocation to achieve that https://kyuubi.apache.org/docs/latest/deployment/spark/dynamic_allocation.html

**Reply by kmsantanu** [reply]
what is the different between  kyuubi.engine.share.level=CONNECTION and kyuubi.engine.share.level=SERVER/GROUP
 will it support on kyuubi server Kyuubi 1.2.0 (git revision a6ab236).
can you suggest which stable kyuubi version we can upgrade in our production and currntly we are using Kyuubi 1.2.0

**Reply by kmsantanu** [reply]
is it required to configure external databse to manage metadata configuration like mysql

**Reply by pan3793** [reply]
The latest stable version is v1.4.1-incubating, you can download it from https://kyuubi.apache.org/releases.html

**Reply by pan3793** [reply]
No, currently Kyuubi does not require MySQL or something to store metadata, maybe need in future.

**Reply by pan3793** [reply]
> what is the different between kyuubi.engine.share.level=CONNECTION and kyuubi.engine.share.level=SERVER/GROUP

You can learn it from https://kyuubi.apache.org/docs/latest/deployment/engine_share_level.html

**Comment by kmsantanu** [comment]
Recently  Kyuubi Support launched query engine asynchronously during opening session .we did not understand can you please provide some examples

**Comment by kmsantanu** [comment]
Kyuubi has a concept names "engine share level", ,is it any dependency with derby database. In production we have seen derby databae is creating inside kyuubi. we are not using any hive query.we are using hive jdbc to run sql .this is the url "jdbc:hive2://system-west01:2181/proddb" we are using..
quetion 1: will derby db created automatically if we connect to bee-line or using url "jdbc:hive2://system-west01:2181/proddb"
quetion 2: is there a performnce issue because of creating derby database for using kyuubi new feature "engine share level",?

**Comment by kmsantanu** [comment]
second quetion: is it required to set both "engine share level" and user level dynamic partion like configuration example kent.spark.dynamicAllocation.maxExecutors=20
"kent.spark.dynamicAllocation.maxExecutors=20"
For example if we set kent.spark.dynamicAllocation.maxExecutors=20 and bob.spark.dynamicAllocation.maxExecutors=10 configuration with dynamic resource allocation then two kyuubi spark engine will start and serve the request.Then what is use of "engine share level" configuration.

can anyone gives the comments please

**Reply by yaooqinn** [reply]
ideally，we shall use the group name that both users belong to. but currently this is not supported yet. the current implementation will let the user come first define the engine settings

**Reply by kmsantanu** [reply]
@[yaooqinn] as your above answer ,I understood that  engine share leve setting to group level is only option and as per kyubi documentation setting of fwe dynamic allocation properties user lavel  that is kent.spark.dynamicAllocation.maxExecutors=20 not working.

**Reply by yaooqinn** [reply]
> @[yaooqinn] as your above answer ,I understood that engine share leve setting to group level is only option and as per kyubi documentation setting of fwe dynamic allocation properties user lavel that is kent.spark.dynamicAllocation.maxExecutors=20 not working.

Sorry, I don't quite understand what you were trying to tell me. Can you make yourself more clear?

**Reply by yaooqinn** [reply]
BTW, keep asking questions irrelevant to the original topic under a `resolved discussion` may disturb the irrelevant people. Would you mind creating a separate one, and making the question clear?

**Reply by zinking** [reply]
@kmsantanu  that's what engine level provides you. you set it to user and then you get to setup two users with different DRA capabilities.

**Comment by kmsantanu** [comment]
can some one clarify above please
